x-bytestring/src/X/Data/ByteString.hs

```diff
@@ -0,0 +1,116 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module X.Data.ByteString (
+    eq1
+  , eq2
+  , eq3
+  , eq4
+  , eq5
+
+  , module Data.ByteString
+  ) where
+
+import           Data.Bits
+import           Data.ByteString
+import           Data.ByteString.Internal (ByteString(..), accursedUnutterablePerformIO)
+import           Data.Word (Word8, Word16, Word32)
+
+import           Foreign.ForeignPtr (withForeignPtr)
+import           Foreign.Storable (peekByteOff)
+
+import           P
+
+
+-- | Fast equality for a 1-byte string.
+eq1 :: Word8 -> ByteString -> Bool
+eq1 x0 (PS fp off len) =
+  let
+    ok =
+      accursedUnutterablePerformIO $! withForeignPtr fp $! \ptr -> do
+        !y0 <- peekByteOff ptr off
+        pure $!
+          x0 == y0
+  in
+    len == 1 && ok
+{-# INLINE eq1 #-}
+
+-- | Fast equality for a 2-byte string.
+eq2 :: Word8 -> Word8 -> ByteString -> Bool
+eq2 x0 x1 (PS fp off len) =
+  let
+    x01 :: Word16
+    !x01 =
+      (fromIntegral x0) .|.
+      (fromIntegral x1 `shiftL` 8)
+
+    ok =
+      accursedUnutterablePerformIO $! withForeignPtr fp $! \ptr -> do
```

Comment:
it's very similar to [this](https://github.com/haskell/bytestring/blob/ce3c6a01cd9e2438358108e3e4b515304b5ce5a6/Data/ByteString.hs#L502-L511) usage, if that helps?